### ISSUE TITLE
Add pinyin on some rarely-used characters

### ISSUE THREAD

**Issue by SleepyAsh0191** [xicri/genshin-langdata#114]
In the text of Liyue, there are also characters that are easily misread or unfamiliar. We call them "rarely-used characters" (生僻字).

Such as 靖妖<ruby>傩<rt>nuó</rt></ruby>舞(Xiao's voice), 天云<ruby>峠<rt>qiǎ</rt></ruby>(Location in Inazuma), and so on. I was wondering if it's possible to add pinyin display on the web page (only for rarely-used characters), just like adding hiragana to reading materials for Japanese children.

As for the data structure in JSON, I am still contemplating.

Note: This is just a suggestion.

**Comment by xicri** [comment]
Sounds great! Let's add pinyin to the translations.

I plan to update dictionary JSON spec, so I will add `.["zh-CN"].pinyin` property so that you can add pinyin. (See #40 for the full JSON example)

However, unfortunately, I'm currently migrating the website from Nuxt 2/Vue to Next.js/React. I think I can start to implement this after the migration, so I may need time.

Regarding to the UI, does it look OK to add pinyin like Hiragana for Japanese?
I expect following UI. (pinyin and hiragana are displayed in smaller text.)

简体中文: 靖妖傩舞 (jìng yāo nuó wǔ)
英语:	Evil conquering!
日语:	靖妖儺舞 (せいようなぶ)

-----

I currently add Chinese pronunciation in katakana in `notes` field of some translations, and I plan to create `kana` field to move them. I plan to display them only in Japanese UI (/ja/*).
By adding pinyin, Japanese UI should look like this:

日本語:	靖妖儺舞 (せいようなぶ)
英語:	Evil conquering!
中国語:	靖妖傩舞 (jìng yāo nuó wǔ / シン ヤオ ヌオ ウー)

I think kana expression is useful for users who has no experience in Chinese, but I had a complaint from a Japanese who has some experience in Chinese because you can never express accurate pronunciation in kana.
If we added pinyins in addition to kana pronunciation, they should be useful for non-native Chinese speakers too.

Thanks for your suggestion!

**Comment by SleepyAsh0191** [comment]
> Regarding to the UI, does it look OK to add pinyin like Hiragana for Japanese?
> I expect following UI. (pinyin and hiragana are displayed in smaller text.)
I think this would perfectly fine.

**Comment by Bill-Haku** [comment]
> only for rarely-used characters

Regarding to this issue, I would prefer this plan. Also, we seldom put pinyin next to words, but just above it. I'd appreciate it if you can try to implement putting pinyin above words, with `<ruby/>` label in HTML.

Regarding to add kana for Chinese words, it would be better to be designed as a section for Japanese users, and applied for words come from Chinese.

**Comment by xicri** [comment]
@Bill-Haku Thanks for the feedback!

>> only for rarely-used characters
> Regarding to this issue, I would prefer this plan.

OK, maybe that's better. Currently, we don't know how many non-native Chinese speakers are interested in pinyin, so it looks better not to add pinyin for them, at least for now.

> Also, we seldom put pinyin next to words, but just above it. I'd appreciate it if you can try to implement putting pinyin above words, with `<ruby/>` label in HTML.

OK, I'll do so.

When I implement the ruby, is the following enough?

A.
![](https://user-images.githubusercontent.com/107226297/232760555-8e71ec5b-e528-497a-81ca-682b8779ee53.png)

```html
<ruby>钟离<rt>zhōng lí</rt></ruby>
```

Or do you think the following is preferable?

B.

![](https://user-images.githubusercontent.com/107226297/232760553-b9522649-76a7-4c29-b487-090ecde82b2d.png)

```html
<ruby>钟<rt>zhōng</rt></ruby><ruby>离<rt>lí</rt></ruby>
```

The first one is easier to implement, but if you want each chunk (zhōng and lí) to be on the corresponding Hanzi (钟 and 离), I need to implement by the second way.

> Regarding to add kana for Chinese words, it would be better to be designed as a section for Japanese users, and applied for words come from Chinese.

In my plan, kana for Chinese words are only displayed in Japanese UI ( https://genshin-dictionary.com/ja/* ).
Do you think it's OK?

**Comment by Bill-Haku** [comment]
> Currently, we don't know how many non-native Chinese speakers are interested in pinyin, so it looks better not to add pinyin for them, at least for now.

I think adding pinyin are basically provided for Chinese users, as Genshin sometimes use not common characters, it will be helpful for them to read the words. Although these common characters can be found in a dictionary, it is hard to type them in if we don't know how to read them in pinyin, which makes it difficult to directly look up the dictionary. Also, words like this is highly likely difficult to be understood, so they can use genshin-dictionary, will just typing the other characters to search them. For foreigner users, it can also be helpful if they study Chinese, but generally effect is limited.

> When I implement the ruby, is the following enough?

I would prefer plan B. Each pinyin should match each character, and there's no the such situation similar to that in Japanese that 2 or more Kanji is corresponded to a set of Kana phrase. Also, adding pinyin is only necessary for one or single character in a few words of all, maybe 10~15% of all the words in genshin-dictionary. Adding pinyin for each character in Chinese is not suggested for texts made for people older than high-grade primary school.

Considering how to implement it, you can use an `Array` to store pinyin, and leave empty for characters which don't need a pinyin. for example, for the word 钟离, if I want to add pinyin for 离 (actually none of the characters needs pinyin here but just an example) :

```
{
    "zhCN": "钟离",
    "pinyin": ["", "lí"],
    ...
}
```

> In my plan, kana for Chinese words are only displayed in Japanese UI ( https://genshin-dictionary.com/ja/* ).

I think it is great!

**Comment by xicri** [comment]
@Bill-Haku Sorry for the late reply. I recently get sleepy at 9 pm (== 8 pm CST) and hard to write a reply when you posted the comment. I'm no longer young... :sweat_smile:

> I would prefer plan B. Each pinyin should match each character,

> Also, adding pinyin is only necessary for one or single character in a few words of all

OK, so if 离 was a rarely used character, you mean like this, right?

![](https://user-images.githubusercontent.com/107226297/232906209-f1544850-0f83-4e78-b546-2e37dd0fff44.png)

```html
钟<ruby>离<rt>lí</rt></ruby>
```

> Considering how to implement it, you can use an Array to store pinyin, and leave empty for characters which don't need a pinyin.

I don't think it is good for long words.
For example, consider 阿斯托洛吉斯·莫娜·梅姬斯图斯 and 图 was a rarely-used character. In this case, too many empty strings are required.

```json
{
    "zhCN": "阿斯托洛吉斯·莫娜·梅姬斯图斯",
    "pinyin": [ "", "", "", "", "", "", "", "", "", "", "", "", "", "tú", "" ],
    ...
}
```

I'm considering the following implementation, but this might make the implementation more complex. Let me know if you find a better one.

```json
{
    "zhCN": "阿斯托洛吉斯·莫娜·梅姬斯图斯",
    "pinyin": [{ "图": "tú" }],
    ...
}
```

**Comment by xicri** [comment]
@Bill-Haku

> I'm considering the following implementation, but this might make the implementation more complex. Let me know if you find a better one.

I further considered how to implement it and found it would not be so difficult by changing the JSON format a little bit:

```json
{
    "zhCN": "阿斯托洛吉斯·莫娜·梅姬斯图斯",
    "pinyins": [{ han: "图", pinyin "tú" }],
    ...
}
```

```javascript
// const word = <above JSON>
const htmls = [];

for (const pinyin of word.pinyins) {
  const html = word.zhCN.replaceAll(pinyin.han, `<ruby>${pinyin.han}<rt>${pinyin.pinyin}</rt></ruby>`);
  htmls.push(html);
}
```

I hear Chinese characters have a single pronunciation for each character unlike Japanese, so I think [`replaceAll`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll) works as expected even if the same rarely-used-characters appears twice.
e.g. 七七 would be "<ruby>七<rt>qī</rt></ruby><ruby>七<rt>qī</rt></ruby>" if 七 was a rarely-used-characters.

---

By the way, how do you think about the property key names `han` and `pinyin` in `"pinyins": [{ han: "图", pinyin "tú" }]`?

Regarding `han`, as you probably know, characters imported from Chinese are called Kanji as a comparison to Kana in Japanese, and in the same way, I guess Chinese characters are called "Han" or "Hanzi" (汉字) as a comparison to pinyin. Is it correct? Do you think using `han` as a property name sounds natural?

In addition, I want to avoid `pinyin` property for `pinyins` because it repeats "pinyin" twice if possible. Replacing `pinyins` 
with `pronunciationZh` can be an idea, but I want to keep the parent property name simple like `kana`.

**Comment by Bill-Haku** [comment]
> Sorry for the late reply. I recently get sleepy at 9 pm (== 8 pm CST) and was hard to write a reply when you posted the comment.

No problem, and also it seems that you wake up very early, which is a healthy lifestyle.

> I further considered how to implement it and found it would not be so difficult by changing the JSON format a little bit:

I like this format. 

> I hear Chinese characters have a single pronunciation for each character unlike Japanese, so I think [replaceAll](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll) works as expected even if the same rarely-used-characters appears twice.

Although a small number of characters have different pronunciations, most of them are common characters so it is not a problem for us. Also, 2 pronunciations of one character appearing in one word is something I am hard to come up with an example, we don't need to worry about it.

> By the way, how do you think about the property key names `han` and `pinyin`

Actually, pinyin is something never being heard of until the western people invented it in the late 19 century. The relationship of Hanzi and pinyin is quilt different from that of Kanji and Kana. Pinyin is simply born to be a tool to pronounce Chinese. In my English learning experience, my teachers always tell us that the English of 汉字 is Chinese characters. Maybe the word of hanzi is not a common usage in English. 

Regarding to the property name, here's my suggestion: `pinyins": [{ "char": "图", "pronounce": "tú" }]`. But I think it should be discussed further.

**Comment by xicri** [comment]
>> I further considered how to implement it and found it would not be so difficult by changing the JSON format a little bit:
>
> I like this format.

>> I hear Chinese characters have a single pronunciation for each character unlike Japanese, so I think [replaceAll](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll) works as expected even if the same rarely-used-characters appears twice.
> 
> Although a small number of characters have different pronunciations, most of them are common characters so it is not a problem for us. Also, 2 pronunciations of one character appearing in one word is something I am hard to come up with an example, we don't need to worry about it.

Thanks, I will start implementation by this way.

> > By the way, how do you think about the property key names `han` and `pinyin`
> 
> Actually, pinyin is something never being heard of until the western people invented it in the late 19 century. The relationship of Hanzi and pinyin is quilt different from that of Kanji and Kana. Pinyin is simply born to be a tool to pronounce Chinese. In my English learning experience, my teachers always tell us that the English of 汉字 is Chinese characters. Maybe the word of hanzi is not a common usage in English.
> 
> Regarding to the property name, here's my suggestion: `pinyins": [{ "char": "图", "pronounce": "tú" }]`. But I think it should be discussed further.

Thanks, `pinyins": [{ "char": "图", "pronounce": "tú" }]` looks nice! Because "pronounce" seems a verb but the noun form "pronunciation" is a bit too long, I will use ["pron"](https://www.merriam-webster.com/dictionary/pron) instead: `pinyins": [{ "char": "图", "pron": "tú" }]`

**Comment by xicri** [comment]
@SleepyAsh0191 Does the above implementation look OK to you?
Let me know if you have any idea when you have time :)

**Comment by Bill-Haku** [comment]
`"pinyins": [{ "char": "图", "pron": "tú" }]` This format also looks good for me!

**Comment by xicri** [comment]
@Bill-Haku @SleepyAsh0191 Can you help me with validating pinyin?
I want to validate pinyin strings in `pron` before the dictionary JSON is built on the CI/CD.

I googled and found [a regex to validate pinyin](https://stackoverflow.com/questions/20736291/regex-for-matching-pinyin), but it is too complex.
Therefore, I will just validate if there are only pinyin characters included in the `pron`. Is the following regex matches the pinyin?

```
/^[a-züāēīōūǖáéíóúǘǎěǐǒǔǚàèìòùǜ]*$/
```

- I wonder if there are any missing characters
- I wonder if I can validate the tone letters used for 四声 with a bit more simple expression like `a-z` rather than listing all the characters
    - I checked the [Unicode characters list](https://en.wikipedia.org/wiki/List_of_Unicode_characters#Latin_Extended-A), but it seems difficult.

**Comment by Bill-Haku** [comment]
There's no missing characters.

There's an informal expression of pinyin 四声, as typing characters with tone is difficult in daily life. For example,  `tú` can written as `tu2`, as ú means the 2nd tone of 四声. Also, you can consider using program code to convert to `tu2` to `tú`. As this is too difficult and complex, I will not tell too much details and things you should pay attention to about this here. 

Otherwise, we'd better just write what it should be.

**Comment by xicri** [comment]
@Bill-Haku Thanks for checking the regex!

> Also, you can consider using program code to convert to tu2 to tú.

OK, I will replace the alphabet + the number (e.g. `tu2`) with a single character with the tone letter (e.g. `tú`) on the build time.

I will ask you for the code review when I have done it.

**Comment by Bill-Haku** [comment]
Let's confirm it: in source json files, we write pinyin in `tu2`, and when building it, transform it into `tú` and display it with tones.

If you are sure to implement it, it would be a lot easier for us when editing the source files. I just want to remind you the priority when adding tone on characters when more than one characters of `aeiouǜ` exist. If `a` exists, add tone on `a` first. If `a` does not exist, add tone on `o` or `e`. `o` and `e` will not appear at the same time. When `i` and `u` appear at the same time, add on the latter character. For example, `liu2` should be `liú`, and `gui4` should be `guì`. `ǜ` always appears alone.

**Comment by SleepyAsh0191** [comment]
> Let's confirm it: in source json files, we write pinyin in `tu2`, and when building it, transform it into `tú` and display it with tones.
> 
> If you are sure to implement it, it would be a lot easier for us when editing the source files. I just want to remind you the priority when adding tone on characters when more than one characters of `aeiouü` exist. If `a` exists, add tone on `a` first. If `a` does not exist, add tone on `o` or `e`. `o` and `e` will not appear at the same time. When `i` and `u` appear at the same time, add on the latter character. For example, `liu2` should be `liú`, and `gui4` should be `guì`. `ü` always appears alone.

use `v` and replace it to `ü` will have a good experience to make

**Comment by Bill-Haku** [comment]
> use `v` and replace it to `ü` will have a good experience to make

Yes, I support it.

**Comment by xicri** [comment]
@Bill-Haku

> Let's confirm it: in source json files, we write pinyin in tu2, and when building it, transform it into tú and display it with tones.

Yes, in my implementation plan, you can write pinyin in the `tu2` form in the source JSON5 files, and just before the dictionary is deployed, `tu2` is transformed to `tú`. End users will see `tú`.
You can also write in the `tú` form in source JSON5 if you prefer.

...However, it sounds more complex than I expected.
I wrote the following code to transform `tu2` form, but do I need more logic to properly convert it?
The following JS just replaces `a1` to `ā`, `a2` to `á`, `a3` to `ǎ`, ...

```js
this.#words.map(word =>
      word.pinyins.map(pinyin => {
        pinyin.pron = pinyin.pron
          .replaceAll("a1", "ā")
          .replaceAll("a2", "á")
          .replaceAll("a3", "ǎ")
          .replaceAll("a4", "à")
          .replaceAll("e1", "ē")
          .replaceAll("e2", "é")
          .replaceAll("e3", "ě")
          .replaceAll("e4", "è")
          .replaceAll("i1", "ī")
          .replaceAll("i2", "í")
          .replaceAll("i3", "ǐ")
          .replaceAll("i4", "ì")
          .replaceAll("o1", "ō")
          .replaceAll("o2", "ó")
          .replaceAll("o3", "ǒ")
          .replaceAll("o4", "ò")
          .replaceAll("u1", "ū")
          .replaceAll("u2", "ú")
          .replaceAll("u3", "ǔ")
          .replaceAll("u4", "ù")
          // TODO transform `v` to `ü`
          .replaceAll("ü1", "ǖ")
          .replaceAll("ü2", "ǘ")
          .replaceAll("ü3", "ǚ")
          .replaceAll("ü4", "ǜ");
        return pinyin;
      })
    );
```

@SleepyAsh0191

> use `v` and replace it to `ü` will have a good experience to make
> 
> e.g: In Chinese IME, to input "女生" (pinyin: nǚ shēng), you only need to type "nvsheng" on the keyboard.
> 
> Because the letter "v" never appear in Chinese Pinyin.

Thanks, I wanted to check how can I handle `ü`.

However, I thought you need to type `v3` to `ǚ`, but in your example, `v` (without `3`) is converted to `ǚ`.
Should I always transform `v` to `ǚ`, or are there more complex rules?

**Comment by Bill-Haku** [comment]
> The following JS just replaces a1 to ā, a2 to á, a3 to ǎ, ...

There's a problem, the number is always written at the end, so this code may node work properly. If you cannot come up with a easy solution, we may consider write the number just behind the correct letter, and you also don't need to consider the priority of letters on which the tone should be added.

> However, I thought you need to type v3 to ǚ, but in your example, v (without 3) is converted to ǚ.
Should I always transform v to ǚ, or are there more complex rules?

In Chinese IME, we don't need to type the tone, and IME will help us select the possible words. So just consider `v` the same as `ü`.

**Comment by xicri** [comment]
@Bill-Haku Thanks, I understand the number is appended to the last of each chunk of pronunciation (e.g. `xi3huan1` → `xǐ huān`, not `xi3hua1n`)

It seems a bit too complex to implement by myself, so I searched npm and found some libraries to transform.

https://www.npmjs.com/package/pinyin-convert
https://www.npmjs.com/package/pinyin-tone-convert

These packages are a bit old, but they are just simple scripts and also have English documents, I will consider using them.

**Comment by Bill-Haku** [comment]
Looking forward to yours good news.

**Comment by SleepyAsh0191** [comment]
Same

**Comment by xicri** [comment]
I decided to use the following package to transform tone numbers (`tu2`) to tone letters (`tú`) because:

- it is relatively well-maintained
    - The last update is a year ago. Not great, but better than the above two packages.
- it seems developed by a native Chinese speaker
- it supports TypeScript's type definition

https://www.npmjs.com/package/pinyin-tone

**Comment by xicri** [comment]
@Bill-Haku @SleepyAsh0191

I have finished the implementation and deployed it to the staging server.
I have added pinyin to 靖妖傩舞 and 天云峠, so can you check if the pinyin is displayed as expected?

https://staging.genshin-dictionary.com/ja/evil-conquering/
https://staging.genshin-dictionary.com/ja/amakumo-peak/

I also added some comments in the code in #128, so can you take a look and check if my implementation is correct?

**Comment by Bill-Haku** [comment]
> https://www.npmjs.com/package/pinyin-tone

This package looks nice. 

> I have added pinyin to 靖妖傩舞 and 天云峠, so can you check if the pinyin is displayed as expected?

I have checked them. They also works nice!

**Comment by xicri** [comment]
@Bill-Haku Thanks for the review!
Now I will merge the PRs. You might encounter some issues when you work on adding more pinyin in the dictionary, but since it does not affect the current production environment, let me fix those issues when you start adding pinyin.

**Comment by xicri** [comment]
@Bill-Haku @SleepyAsh0191 Now I merged the PRs (xicri/genshin-dictionary#88 and #114) and you can start adding pinyin!
